### ISSUE TITLE
Matching generic return type of `SupportsGetItem`

### ISSUE THREAD

**Issue by cmp0xff** [python/mypy#21731]
## Bug Report

`mypy` does not match the generic return type of `SupportsGetItem` properly.

This was proposed in https://github.com/astral-sh/ty/issues/3988#issuecomment-4960290236.

## To Reproduce

https://mypy-play.net/?gist=b8bb3085f195314a6c7c89ed2bdd2576

```python
from __future__ import annotations

from typing import TYPE_CHECKING, Any, TypeVar

if TYPE_CHECKING:
    from _typeshed import SupportsGetItem

class Series: ...

class DataFrame:
    def __getitem__(self, key: object) -> Series:
        return Series()

T = TypeVar("T")

def f(x: SupportsGetItem[Any, T]) -> T:
    if isinstance(x, DataFrame):
        # return Series()
        return 1  # false error-free
    else:
        return x[0]

```

## Expected Behaviour

Report an error about return type

## Actual Behaviour

No error

## Your Environment

`mypy` playground

**Comment by Jbansal2** [comment]
Hi! I'd like to investigate and work on fixing this issue. If no one else is already working on it, could you please assign it to me?

**Comment by hauntsaninja** [comment]
(Note if you enable `--warn-unreachable` you will get an error here. Obviously there is a bug and the code is still reachable though)

**Comment by cmp0xff** [comment]
Thank you @hauntsaninja ! Ideally when the bug is fixed, `return Series()` can also be error-free. Currently none of `pyright`, `pyrefly` or `ty` supports this feature, see microsoft/pyright#11544 facebook/pyrefly#4135 astral-sh/ty#3988.
